compaction/compaction_garbage_collector.hh

```diff
@@ -22,7 +22,19 @@ using can_gc_fn = std::function<bool(tombstone, is_shadowable)>;
 extern can_gc_fn always_gc;
 extern can_gc_fn never_gc;
 
-using max_purgeable_fn = std::function<api::timestamp_type(const dht::decorated_key&, is_shadowable)>;
+struct max_purgeable {
+    enum class timestamp_source {
+        memtable_possibly_shadowing_data,
+        other_sstables_possibly_shadowing_data
+    };
+
+    operator bool() const { return timestamp != api::missing_timestamp; }
+
+    api::timestamp_type timestamp { api::missing_timestamp };
+    std::optional<timestamp_source> source;
+};
+
+using max_purgeable_fn = std::function<max_purgeable(const dht::decorated_key&, is_shadowable)>;
```

Comment:
@denesb could this have an impact on the read path?

Reply:
We calculate max purgeable once per partition for a reason. I am not sure we're increasing number of times here, but if we do, it's certainly a regression.

Reply:
This changes the return value from bool to some struct. I don't think will have any visible impact.
Like Raphael said, we call this method once per each partition (maybe twice, if there are shadowable tombstones in the partition too).

Reply:
I do not increase the frequency the function is called. Only the return type is changed.

Reply:
Is the complexity of calculating the return type materially increased? If not, ok.